collect_app/src/main/java/org/odk/collect/android/storage/StoragePathProvider.java

```diff
@@ -1,29 +1,78 @@
+        externalFilesDirPath = Collect.getInstance().getExternalFilesDir(null).getAbsolutePath();
+    }
+
+    public StoragePathProvider(CurrentProjectProvider currentProjectProvider, String externalFilesDirPath) {
+        this.currentProjectProvider = currentProjectProvider;
+        this.externalFilesDirPath = externalFilesDirPath;
+    }
+
+    public String[] getOdkRootDirPaths() {
+        return new String[]{
+                getOdkDirPath(StorageSubdirectory.PROJECTS)
+        };
+    }
+
+    public String[] getProjectDirPaths(Project.Saved project) {
         return new String[]{
-                getOdkDirPath(StorageSubdirectory.FORMS),
-                getOdkDirPath(StorageSubdirectory.INSTANCES),
-                getOdkDirPath(StorageSubdirectory.CACHE),
-                getOdkDirPath(StorageSubdirectory.METADATA),
-                getOdkDirPath(StorageSubdirectory.LAYERS)
+                getOdkDirPath(StorageSubdirectory.FORMS, project),
+                getOdkDirPath(StorageSubdirectory.INSTANCES, project),
+                getOdkDirPath(StorageSubdirectory.CACHE, project),
+                getOdkDirPath(StorageSubdirectory.METADATA, project),
+                getOdkDirPath(StorageSubdirectory.LAYERS, project)
         };
     }
 
     public String getOdkRootDirPath() {
-        File odkDirPath = Collect.getInstance().getExternalFilesDir(null);
-        return odkDirPath != null
-                ? odkDirPath.getAbsolutePath()
-                : "";
+        return externalFilesDirPath;
+    }
+
+    public String getProjectRootDirPath() {
+        return getProjectRootDirPath(null);
+    }
+
+    public String getProjectRootDirPath(Project.Saved project) {
+        if (project == null) {
+            String projectId = currentProjectProvider.getCurrentProject().getUuid();
+            return getOdkDirPath(StorageSubdirectory.PROJECTS) + File.separator + projectId;
+        } else {
+            return getOdkDirPath(StorageSubdirectory.PROJECTS) + File.separator + project.getUuid();
+        }
     }
 
     public String getOdkDirPath(StorageSubdirectory subdirectory) {
-        return getOdkRootDirPath() + File.separator + subdirectory.getDirectoryName();
+        return getOdkDirPath(subdirectory, null);
+    }
+
+    public String getOdkDirPath(StorageSubdirectory subdirectory, Project.Saved project) {
+        switch (subdirectory) {
+            case FORMS:
+            case INSTANCES:
+            case CACHE:
+            case METADATA:
+            case LAYERS:
+                return getProjectRootDirPath(project) + File.separator + subdirectory.getDirectoryName();
+            case SETTINGS:
+            case PROJECTS:
+                return getOdkRootDirPath() + File.separator + subdirectory.getDirectoryName();
+            default:
+                throw new IllegalStateException("Unexpected value: " + subdirectory);
+        }
     }
 
     public String getCustomSplashScreenImagePath() {
```

Comment:
I think we need to have a separate issue for this case. If we store all splash screen images in the same place and using the same name it's no possible to set different images for different projects. I'm not even sure what the expected behavior is but definitely it's something we should think through.

Reply:
~~Hmmmm. I think because of the new [splash screen API](https://developer.android.com/about/versions/12/features/splash-screen) (which looks like it only supports hardcoded splash icons) we probably want to deprecate the custom splash screen image sometime soon.~~

There's an [issue for just removing the splash screen configuration](https://github.com/getodk/collect/issues/254) that we probably want to schedule within the next couple of releases. @lognaturel do you think we should go ahead and plan this for 2021.3?